File: jfix-zookeeper/src/test/java/ru/fix/zookeeper/transactional/ZkTransactionTest.java
package ru.fix.zookeeper.transactional;

import org.apache.curator.framework.CuratorFramework;
import org.apache.zookeeper.KeeperException;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ru.fix.zookeeper.testing.ZKTestingServer;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.*;

import static org.junit.jupiter.api.Assertions.*;


public class ZkTransactionTest {

    private ZKTestingServer zkServer;

    @BeforeEach
    public void startZkServer() throws Exception {
        zkServer = new ZKTestingServer()
                .withCloseOnJvmShutdown();
        zkServer.start();
    }

    @AfterEach
    public void stopZkServer(){
        zkServer.close();
    }


    @Test
    public void testDeletePathWithChildrenIfNeeded() throws Exception {
        zkServer.getClient().create().creatingParentsIfNeeded().forPath("/1/2/3/4/5/6/7");

        ZkTransaction.createTransaction(zkServer.getClient())
                .deletePathWithChildrenIfNeeded("/1/2/3")
                .deletePathWithChildrenIfNeeded("/1/2")
                .commit();

        assertNull(zkServer.getClient().checkExists().forPath("/1/2"));
        assertNotNull(zkServer.getClient().checkExists().forPath("/1"));
    }

    @Test
    public void testCreatePathWithParentsIfNeeded() throws Exception {
        zkServer.getClient().create().creatingParentsIfNeeded().forPath("/2/03");

        ZkTransaction.createTransaction(zkServer.getClient())
                .createPathWithParentsIfNeeded("/1/01/001")
                .checkPath("/1/01/001")
                .createPathWithParentsIfNeeded("/1/01/001/0001/00001")
                .createPathWithParentsIfNeeded("/2/03/003")
                .commit();

        assertNotNull(zkServer.getClient().checkExists().forPath("/1/01/001"));
        assertNotNull(zkServer.getClient().checkExists().forPath("/1/01/001/0001/00001"));
        assertNotNull(zkServer.getClient().checkExists().forPath("/2/03/003"));
    }

    @Test
    public void testCreatePathWithParentsIfNeeded_ForExistingNode() throws Exception {
        zkServer.getClient().create().creatingParentsIfNeeded().forPath("/2/03");
        assertThrows(
                KeeperException.NodeExistsException.class,
                () -> ZkTransaction.createTransaction(zkServer.getClient())
                        .createPathWithParentsIfNeeded("/2/03")
                        .commit()
        );
    }

    @Test
    public void testCheckPathWithVersion() throws Exception {
        ZkTransaction.createTransaction(zkServer.getClient())
                .createPathWithParentsIfNeeded("/1/01/001")
                .checkPath("/1/01/001")
                .checkPathWithVersion("/1/01/001", 0)
                .setData("/1/01/001", new byte[]{101})
                .checkPathWithVersion("/1/01/001", 1)
                .commit();

        assertNotNull(zkServer.getClient().checkExists().forPath("/1/01/001"));
        assertArrayEquals(new byte[]{101}, zkServer.getClient().getData().forPath("/1/01/001"));
    }

    @Test
    public void testCheckPathWithIncorrectVersion() throws Exception {
        assertThrows(
                KeeperException.BadVersionException.class,
                () -> ZkTransaction.createTransaction(zkServer.getClient())
                        .createPathWithParentsIfNeeded("/1/01/001")
                        .checkPath("/1/01/001")
                        .checkPathWithVersion("/1/01/001", 0)
                        .setData("/1/01/001", new byte[]{101})
                        .checkPathWithVersion("/1/01/001", 2)
                        .commit()
        );
    }

    /**
     * Test case for unsupported delete and create operations mix
     */
    @Test
    public void testMixedCreateDelete_Failure() throws Exception {
        zkServer.getClient().create()
                .creatingParentsIfNeeded()
                .forPath("/1/2/3/4/5");

        assertThrows(
                KeeperException.NoNodeException.class,
                () -> ZkTransaction.createTransaction(zkServer.getClient())
                        .deletePathWithChildrenIfNeeded("/1/2/3")
                        .createPathWithParentsIfNeeded("/1/2/3/4")
                        .commit()
        );
    }


    @Test
    public void check_version_WHEN_multiply_updating_THEN_only_one_succeed() throws Exception {
        String lockPath = "/lock";
        CuratorFramework curator = zkServer.getClient();
        curator.create().creatingParentsIfNeeded().forPath(lockPath);

        Set<String> expectedNodes = new CopyOnWriteArraySet<>();
        expectedNodes.add("lock");

        int quantityOfTransactions = 10;
        CountDownLatch countDownLatch = new CountDownLatch(quantityOfTransactions);
        List<CompletableFuture<Void>> futures = new ArrayList<>(quantityOfTransactions);
        ExecutorService executor = Executors.newFixedThreadPool(quantityOfTransactions);

        for (int i = 0; i < quantityOfTransactions; i++) {
            final int transactionNumber = i;
            futures.add(CompletableFuture.runAsync(() -> {
                try {
                    ZkTransaction.tryCommit(zkServer.getClient(), 1, transaction -> {
                        transaction.checkAndUpdateVersion(lockPath);
                        transaction.createPath("/" + transactionNumber);
                        countDownLatch.countDown();
                        countDownLatch.await();
                    });
                    expectedNodes.add(String.valueOf(transactionNumber));
                } catch (Exception ignored) {
                }
            }, executor));
        }
        CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).join();
        executor.shutdown();

        assertEquals(expectedNodes.size(), 2);
        assertEquals(expectedNodes, new HashSet<>(curator.getChildren().forPath("/")));
    }

}

File: jfix-zookeeper/src/test/kotlin/ru/fix/zookeeper/transactional/ZkTransactionTest.kt
package ru.fix.zookeeper.transactional

import org.apache.curator.framework.recipes.cache.ChildData
import org.apache.curator.framework.recipes.cache.CuratorCache
import org.apache.curator.framework.recipes.cache.CuratorCacheListener
import org.apache.logging.log4j.kotlin.Logging
import org.apache.zookeeper.KeeperException
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import ru.fix.zookeeper.testing.ZKTestingServer
import java.lang.Thread.sleep

import java.util.ArrayList
import java.util.concurrent.*

import java.util.concurrent.atomic.AtomicBoolean


class ZkTransactionTest {
    companion object : Logging

    lateinit var zkServer: ZKTestingServer

    @BeforeEach
    fun startZkServer() {
        zkServer = ZKTestingServer()
                .withCloseOnJvmShutdown();
        zkServer.start();
    }

    @AfterEach
    fun stopZkServer() {
        zkServer.close();
    }


    @Test
    fun deletePathWithChildrenIfNeeded() {
        zkServer.client.create().creatingParentsIfNeeded().forPath("/1/2/3/4/5/6/7");

        ZkTransaction.createTransaction(zkServer.client)
                .deletePathWithChildrenIfNeeded("/1/2/3")
                .deletePathWithChildrenIfNeeded("/1/2")
                .commit();

        assertNull(zkServer.client.checkExists().forPath("/1/2"));
        assertNotNull(zkServer.client.checkExists().forPath("/1"));
    }

    @Test
    fun createPathWithParentsIfNeeded() {
        zkServer.client.create().creatingParentsIfNeeded().forPath("/2/03");

        ZkTransaction.createTransaction(zkServer.client)
                .createPathWithParentsIfNeeded("/1/01/001")
                .checkPath("/1/01/001")
                .createPathWithParentsIfNeeded("/1/01/001/0001/00001")
                .createPathWithParentsIfNeeded("/2/03/003")
                .commit();

        assertNotNull(zkServer.client.checkExists().forPath("/1/01/001"));
        assertNotNull(zkServer.client.checkExists().forPath("/1/01/001/0001/00001"));
        assertNotNull(zkServer.client.checkExists().forPath("/2/03/003"));
    }

    @Test
    fun `createPathWithParentsIfNeeded throws NodeExistsException if node already exist`() {
        zkServer.client.create().creatingParentsIfNeeded().forPath("/2/03");
        assertThrows<KeeperException.NodeExistsException> {
            ZkTransaction.createTransaction(zkServer.client)
                    .createPathWithParentsIfNeeded("/2/03")
                    .commit()
        }
    }

    @Test
    fun checkPathWithVersion() {
        ZkTransaction.createTransaction(zkServer.client)
                .createPathWithParentsIfNeeded("/1/01/001")
                .checkPath("/1/01/001")
                .checkPathWithVersion("/1/01/001", 0)
                .setData("/1/01/001", byteArrayOf(101))
                .checkPathWithVersion("/1/01/001", 1)
                .commit()

        assertNotNull(zkServer.client.checkExists().forPath("/1/01/001"))
        assertArrayEquals(byteArrayOf(101), zkServer.client.getData().forPath("/1/01/001"))
    }

    @Test
    fun checkPathWithIncorrectVersion() {
        assertThrows<KeeperException.BadVersionException> {
            ZkTransaction.createTransaction(zkServer.client)
                    .createPathWithParentsIfNeeded("/1/01/001")
                    .checkPath("/1/01/001")
                    .checkPathWithVersion("/1/01/001", 0)
                    .setData("/1/01/001", byteArrayOf(101))
                    .checkPathWithVersion("/1/01/001", 2)
                    .commit()
        }
    }

    /**
     * Test case for unsupported delete and create operations mix
     */
    @Test
    fun `Mixed create or delete with parents or children fails`() {
        zkServer.client.create()
                .creatingParentsIfNeeded()
                .forPath("/1/2/3/4/5");

        assertThrows<KeeperException.NoNodeException> {
            ZkTransaction.createTransaction(zkServer.client)
                    .deletePathWithChildrenIfNeeded("/1/2/3")
                    .createPathWithParentsIfNeeded("/1/2/3/4")
                    .commit()
        }
    }


    @Test
    fun `with checkVersion only one of multiple updating transactions succeed`() {
        val lockPath = "/lock";
        val curator = zkServer.client;
        curator.create().creatingParentsIfNeeded().forPath(lockPath);

        val expectedNodes = CopyOnWriteArraySet<String>();
        expectedNodes.add("lock");

        val quantityOfTransactions = 10;
        val countDownLatch = CountDownLatch(quantityOfTransactions);
        val futures = ArrayList<CompletableFuture<Void>>(quantityOfTransactions);
        val executor = Executors.newFixedThreadPool(quantityOfTransactions);

        for (transactionNumber in (0 until quantityOfTransactions)) {
            futures.add(CompletableFuture.runAsync(Runnable {
                try {
                    ZkTransaction.tryCommit(zkServer.client, 1) { transaction ->
                        transaction.readVersionThenCheckAndUpdateInTransaction(lockPath);
                        transaction.createPath("/$transactionNumber");
                        countDownLatch.countDown();
                        countDownLatch.await();
                    };
                    expectedNodes.add(transactionNumber.toString());
                } catch (ignored: Exception) {
                }
            }, executor))
        }
        CompletableFuture.allOf(*futures.toTypedArray()).join();
        executor.shutdown();

        assertEquals(expectedNodes.size, 2);
        assertEquals(expectedNodes, curator.children.forPath("/").toHashSet());
    }

    @Test
    fun `transaction with only changeAndUpdateVersionOnChange does nothing`() {
        val curator = zkServer.client
        curator.create().creatingParentsIfNeeded().forPath("/version")

        val changeInZkOccurred = AtomicBoolean()
        val zkListener = CuratorCache.build(curator, "/")
        zkListener.listenable().addListener(CuratorCacheListener { type: CuratorCacheListener.Type,
                                                                   oldData: ChildData?,
                                                                   newData: ChildData? ->
            logger.info("Change $type, ${oldData?.path}, ${newData?.path}")
            changeInZkOccurred.set(true)
        })
        zkListener.start()
        sleep(1000)

        changeInZkOccurred.set(false)
        ZkTransaction.tryCommit(curator, 1) { tx ->
            tx.readVersionThenCheckAndUpdateInTransactionIfItMutatesZkState("/version")
        }
        sleep(1000)
        assertFalse(changeInZkOccurred.get())


        changeInZkOccurred.set(false)
        ZkTransaction.tryCommit(curator, 1){ tx ->
            tx.readVersionThenCheckAndUpdateInTransactionIfItMutatesZkState("/version")
            tx.createPath("/new-path")
        }
        await().until { changeInZkOccurred.get() }

        changeInZkOccurred.set(false)
        ZkTransaction.tryCommit(curator, 1) { tx ->
            tx.readVersionThenCheckAndUpdateInTransactionIfItMutatesZkState("/version")
            tx.setData("/new-path", byteArrayOf(42))
        }
        await().until{ changeInZkOccurred.get() }


        changeInZkOccurred.set(false)
        ZkTransaction.tryCommit(curator, 1) { tx ->
            tx.readVersionThenCheckAndUpdateInTransactionIfItMutatesZkState("/version")
            assertNotNull(curator.checkExists().forPath("/new-path"))
        }
        sleep(1000)
        assertFalse(changeInZkOccurred.get())


        changeInZkOccurred.set(false)
        ZkTransaction.tryCommit(curator, 1){ tx ->
            tx.readVersionThenCheckAndUpdateInTransactionIfItMutatesZkState("/version")
            tx.deletePath("/new-path")
        }
        await().until { changeInZkOccurred.get() }
    }
}